node/network/src/protocol.rs

```diff
@@ -104,19 +123,29 @@ impl Protocol {
                     }
                 }
             }
-            Message::BlockResponse(_request_id, blocks) => {
-                forward_msg(
-                    self.inc_block_tx.clone(),
-                    (peer_id, blocks),
-                );
+            Message::BlockResponse(request_id, blocks) => {
+                match self.requests.read().expect(POISONED_LOCK_ERR).get(&request_id) {
+                    Some(RequestType::BlockFetchRequest(start, end)) => {
+                        let from_index = blocks[0].0.index();
+                        let til_index = blocks[blocks.len() - 1].0.index();
+                        if *start != from_index || *end != til_index {
+                            self.peer_manager.ban_peer(&peer_id);
+                        }
+                    }
+                    _ => self.peer_manager.ban_peer(&peer_id),
+                }
+                forward_msg(self.inc_block_tx.clone(), (peer_id, blocks));
```

Comment:
this request should be removed afterwards (more pointers to helper).

Reply:
It seems that it should only be removed when the check passes. Otherwise an attacker can send some garbage and make you remove some request id.